File: android/src/main/java/ai/xmax/reactnative/XmaxImageManager.kt
package ai.xmax.reactnative

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Matrix
import android.graphics.Paint
import android.graphics.RectF
import android.media.ExifInterface
import android.net.Uri
import com.facebook.react.bridge.Promise
import org.json.JSONObject
import java.io.File
import java.io.InputStream
import java.util.UUID
import java.util.concurrent.Executors
import kotlin.math.max

/** Normalizes local images in private cache without moving pixel buffers through JS. */
internal class XmaxImageManager(private val context: Context) {
  private val executor = Executors.newSingleThreadExecutor()
  private val directory = File(context.cacheDir, "xmax-images")
  private val owned = mutableSetOf<File>()
  private var invalidated = false

  private fun open(value: String): InputStream {
    val uri = if (value.startsWith("/")) Uri.fromFile(File(value)) else Uri.parse(value)
    require(uri.scheme == "file" || uri.scheme == "content") { "A local image URI is required" }
    return requireNotNull(context.contentResolver.openInputStream(uri)) { "Unable to open image" }
  }

  private fun orientation(value: String): Int = open(value).use {
    try { ExifInterface(it).getAttributeInt(ExifInterface.TAG_ORIENTATION, ExifInterface.ORIENTATION_NORMAL) }
    catch (_: java.io.IOException) { ExifInterface.ORIENTATION_NORMAL }
  }

  private fun bounds(value: String): BitmapFactory.Options {
    val options = BitmapFactory.Options().apply { inJustDecodeBounds = true }
    open(value).use { BitmapFactory.decodeStream(it, null, options) }
    require(options.outWidth > 0 && options.outHeight > 0) { "Unable to read image dimensions" }
    return options
  }

  fun info(value: String, promise: Promise) {
    executor.execute {
      try {
        val bounds = bounds(value)
        val rotated = orientation(value) in 5..8
        promise.resolve(JSONObject()
          .put("width", if (rotated) bounds.outHeight else bounds.outWidth)
          .put("height", if (rotated) bounds.outWidth else bounds.outHeight).toString())
      } catch (error: Exception) { promise.reject("MEDIA_ERROR", error) }
    }
  }

  fun prepare(value: String, width: Double, height: Double, promise: Promise) {
    executor.execute {
      var decoded: Bitmap? = null
      var oriented: Bitmap? = null
      var output: Bitmap? = null
      var file: File? = null
      try {
        // Model buckets and pixel bounds are resolved by MediaService before preparation.
        require(width.isFinite() && height.isFinite() && width > 0 && height > 0 &&
          width % 1 == 0.0 && height % 1 == 0.0) { "Invalid prepared image dimensions" }
        val bounds = bounds(value)
        val options = BitmapFactory.Options().apply {
          inPreferredConfig = Bitmap.Config.ARGB_8888
          inSampleSize = 1
          while (max(bounds.outWidth, bounds.outHeight) / inSampleSize > 4096) inSampleSize *= 2
        }
        decoded = open(value).use { BitmapFactory.decodeStream(it, null, options) }
          ?: error("Unable to decode image")
        val transform = Matrix().apply {
          when (orientation(value)) {
            2 -> setScale(-1f, 1f)
            3 -> setRotate(180f)
            4 -> setScale(1f, -1f)
            5 -> { setRotate(90f); postScale(-1f, 1f) }
            6 -> setRotate(90f)
            7 -> { setRotate(-90f); postScale(-1f, 1f) }
            8 -> setRotate(-90f)
          }
        }
        oriented = Bitmap.createBitmap(decoded, 0, 0, decoded.width, decoded.height, transform, true)
        output = Bitmap.createBitmap(width.toInt(), height.toInt(), Bitmap.Config.ARGB_8888)
        val scale = max(width / oriented.width, height / oriented.height).toFloat()
        val left = (width.toFloat() - oriented.width * scale) / 2
        val top = (height.toFloat() - oriented.height * scale) / 2
        Canvas(output).apply {
          drawColor(Color.BLACK)
          drawBitmap(oriented, null, RectF(left, top, left + oriented.width * scale, top + oriented.height * scale), Paint(Paint.FILTER_BITMAP_FLAG))
        }
        check(directory.isDirectory || directory.mkdirs()) { "Unable to create image cache" }
        file = File(directory, "${UUID.randomUUID()}.jpg")
        file.outputStream().use { check(output.compress(Bitmap.CompressFormat.JPEG, 95, it)) { "Unable to save image" } }
        synchronized(this) {
          check(!invalidated) { "Image runtime has been released" }
          owned.add(file)
        }
        promise.resolve(Uri.fromFile(file).toString())
      } catch (error: Exception) {
        file?.delete()
        promise.reject("MEDIA_ERROR", error)
      } catch (error: OutOfMemoryError) {
        file?.delete()
        promise.reject("MEDIA_ERROR", "Insufficient memory to prepare this image", error)
      } finally {
        output?.recycle()
        if (oriented !== decoded) oriented?.recycle()
        decoded?.recycle()
      }
    }
  }

  fun remove(value: String, promise: Promise) {
    executor.execute {
      try {
        val file = File(requireNotNull(Uri.parse(value).path)).canonicalFile
        require(file.parentFile == directory.canonicalFile && file.extension == "jpg") { "Image is outside SDK cache" }
        check(!file.exists() || file.delete()) { "Unable to remove prepared image" }
        synchronized(this) { owned.remove(file) }
        promise.resolve(null)
      } catch (error: Exception) { promise.reject("MEDIA_ERROR", error) }
    }
  }

  /** Runs after engine destruction; queued preparation cannot retain new files. */
  @Synchronized fun invalidate() {
    invalidated = true
    owned.forEach { it.delete() }
    owned.clear()
    executor.shutdown()
  }
}

File: ios/XmaxImageManager.swift
import Foundation
import ImageIO
@preconcurrency import UIKit

/// Decodes local images and owns their prepared cache files; pixels never cross JS.
///
/// Decoding and disk I/O run on worker queues. The lock protects cache ownership
/// against invalidation, so a late preparation cannot leave an orphaned file.
final class XmaxImageManager: @unchecked Sendable {
  /// Protects owned and invalidated; image decoding does not hold this lock.
  private let gate = NSRecursiveLock()
  private var owned = Set<URL>()
  private var invalidated = false

  /// The only directory where this component may create or remove prepared images.
  static var directory: URL {
    FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask)[0]
      .appendingPathComponent("xmax-images", isDirectory: true)
  }

  /// Accepts local file URLs and absolute paths; remote URLs are rejected.
  private static func localURL(_ value: String) -> URL? {
    let url = value.hasPrefix("/") ? URL(fileURLWithPath: value) : URL(string: value)

    return url?.isFileURL == true ? url : nil
  }

  /// Opens ImageIO without eagerly caching the original full-resolution pixels.
  private static func source(_ value: String) -> CGImageSource? {
    guard let url = localURL(value) else {
      return nil
    }

    return CGImageSourceCreateWithURL(
      url as CFURL,
      [kCGImageSourceShouldCache: false] as CFDictionary
    )
  }

  /// Reads oriented dimensions without fully decoding the original image.
  func info(_ fileURL: String, promise: XmaxNativePromise) {
    DispatchQueue.global(qos: .userInitiated).async {
      autoreleasepool {
        guard let source = Self.source(fileURL),
          let properties = CGImageSourceCopyPropertiesAtIndex(source, 0, nil) as? [CFString: Any],
          let width = properties[kCGImagePropertyPixelWidth] as? Int,
          let height = properties[kCGImagePropertyPixelHeight] as? Int,
          width > 0, height > 0
        else {
          promise.reject("Unable to read local image dimensions")
          return
        }

        let orientation = properties[kCGImagePropertyOrientation] as? Int ?? 1
        let rotated = (5...8).contains(orientation)
        let size = ["width": rotated ? height : width, "height": rotated ? width : height]

        do {
          let data = try JSONSerialization.data(withJSONObject: size)
          promise.resolve(String(decoding: data, as: UTF8.self))
        } catch {
          promise.reject("Unable to read local image dimensions", error: error)
        }
      }
    }
  }

  /// Applies EXIF orientation and the requested centered crop to a private JPEG.
  func prepare(_ fileURL: String, width: Double, height: Double, promise: XmaxNativePromise) {
    DispatchQueue.global(qos: .userInitiated).async {
      autoreleasepool {
        // Model buckets and pixel bounds are resolved by MediaService before preparation.
        guard width.isFinite, height.isFinite, width > 0, height > 0,
          width.rounded(.down) == width, height.rounded(.down) == height
        else {
          promise.reject("Invalid prepared image dimensions")
          return
        }

        guard let source = Self.source(fileURL) else {
          promise.reject("Unable to open local image")
          return
        }

        let options: [CFString: Any] = [
          kCGImageSourceCreateThumbnailFromImageAlways: true,
          kCGImageSourceCreateThumbnailWithTransform: true,
          kCGImageSourceShouldCacheImmediately: true,
          kCGImageSourceThumbnailMaxPixelSize: 4096,
        ]
        guard let decoded = CGImageSourceCreateThumbnailAtIndex(source, 0, options as CFDictionary)
        else {
          promise.reject("Unable to decode local image")
          return
        }

        let image = UIImage(cgImage: decoded)
        let scale = max(width / image.size.width, height / image.size.height)
        let target = CGRect(
          x: (width - image.size.width * scale) / 2,
          y: (height - image.size.height * scale) / 2,
          width: image.size.width * scale,
          height: image.size.height * scale
        )

        let format = UIGraphicsImageRendererFormat.default()
        format.scale = 1
        format.opaque = true

        let renderer = UIGraphicsImageRenderer(
          size: CGSize(width: width, height: height),
          format: format
        )
        let data = renderer.jpegData(withCompressionQuality: 0.95) { context in
          UIColor.black.setFill()
          context.fill(CGRect(x: 0, y: 0, width: width, height: height))
          image.draw(in: target)
        }

        let output = Self.directory.appendingPathComponent(UUID().uuidString + ".jpg")

        do {
          try FileManager.default.createDirectory(
            at: Self.directory,
            withIntermediateDirectories: true
          )
          try data.write(to: output, options: .atomic)
        } catch {
          try? FileManager.default.removeItem(at: output)
          promise.reject("Unable to save prepared image", error: error)
          return
        }

        self.gate.xmaxWithLock {
          guard !self.invalidated else {
            try? FileManager.default.removeItem(at: output)
            promise.reject("Image runtime has been released")
            return
          }

          self.owned.insert(output)

          promise.resolve(output.absoluteString)
        }
      }
    }
  }

  /// Removes only prepared SDK JPEGs; a missing file is already considered released.
  func remove(_ fileURL: String, promise: XmaxNativePromise) {
    DispatchQueue.global(qos: .utility).async {
      guard let url = Self.localURL(fileURL)?.resolvingSymlinksInPath(),
        url.deletingLastPathComponent().path == Self.directory.resolvingSymlinksInPath().path,
        url.pathExtension == "jpg"
      else {
        promise.reject("Refusing to remove an image outside SDK cache")
        return
      }

      do {
        try FileManager.default.removeItem(at: url)
      } catch let error as NSError
        where error.domain == NSCocoaErrorDomain && error.code == NSFileNoSuchFileError
      {
        // Removal is idempotent.
      } catch {
        promise.reject("Unable to remove prepared image", error: error)
        return
      }

      self.gate.xmaxWithLock {
        if let original = Self.localURL(fileURL) {
          self.owned.remove(original)
        }
      }

      promise.resolve()
    }
  }

  /// Runs after engine destruction; late preparation deletes its own result.
  func invalidate() {
    gate.xmaxWithLock {
      invalidated = true

      for url in owned {
        try? FileManager.default.removeItem(at: url)
      }

      owned.removeAll()
    }
  }
}
